File: 1000-1200/1091. Shortest Path in Binary Matrix/solution.cpp
class Solution {
private:
    int n;
    int next[8][2] = {{1,0},{-1,0},{0,1},{0,-1},{1,1},{1,-1},{-1,1},{-1,-1}};
    
public:
    int shortestPathBinaryMatrix(vector<vector<int>>& grid) {
        n = grid.size();
        if (grid[0][0] == 1 || grid[n-1][n-1] == 1) return -1;
        if (n == 1) return 1;
        
        queue<pair<int,int>> qu; 
        qu.emplace(0,0);
        grid[0][0] = 1;
        int len = 1;
        while (!qu.empty()) {
            int size = qu.size();
            len++;
            while (size--) {
                auto[i,j] = qu.front(); qu.pop();
                for(int k=0; k<8; k++) {
                    int p = i + next[k][0], q = j + next[k][1];
                    if (p == n-1 && q == n-1) return len;
                    if (!isValid(grid,p,q)) continue;
                    grid[p][q] = 1;
                    qu.emplace(p,q);
                }
            }
        }
        return -1;
    }
    
    bool isValid(vector<vector<int>>& grid, int i, int j) {
        return i>=0 
            && i<n
            && j>=0
            && j<n
            && grid[i][j] == 0;
    }
};

File: 1000-1200/1091. Shortest Path in Binary Matrix/solution.js
var shortestPathBinaryMatrix = function(grid) {
    let i=0
    let j=0
    if (!isValid(grid,i,j)) return -1
    let steps = 1;
    let queue = [[i,j,steps]]
    grid[0][0] =1

    while (queue.length) {
        [i,j,steps] = queue.pop()
        if (i===grid.length-1 && j===grid[i].length-1) return steps

        //eight directional
        let actionSpace = [[i+1,j+1],[i,j+1],[i+1,j],[i-1,j+1],[i+1,j-1],[i-1,j],[i,j-1],[i-1,j-1]]
        for ([i,j] of actionSpace) {
            if (isValid(grid,i,j)) {
                grid[i][j] =1
                queue.unshift([i,j,steps+1])
            }
        }
    }
    return -1
};

let isValid = (grid,i,j) => {
    return i>=0 &&
        i<grid.length &&
        j>=0 &&
        j<grid[i].length &&
        grid[i][j] === 0
}
